test/lib/code_corps/github/api/api_test.exs

```diff
@@ -4,87 +4,122 @@ defmodule CodeCorps.GitHub.APITest do
-      end)
+    test "handles a 200..299 response" do
+      {:ok, response} = API.request(:get, "/200", %{}, [], [])
+      assert response == %{"bar" => "baz"}
+    end
 
-      {:error, response} = API.request(:get, url_for("/foo"), [{"foo", "bar"}], @body, [:with_body])
+    test "handles a decode error for a 200..299 response" do
+      {:error, response} = API.request(:get, "/200-bad", %{}, [], [])
       assert response == HTTPClientError.new([reason: :body_decoding_error])
     end
 
-    test "handles a 404 response", %{bypass: bypass} do
-      Bypass.expect(bypass, "GET", "/foo", fn %Conn{req_headers: headers} = conn ->
-        assert {"foo", "bar"} in headers
-        conn |> Conn.resp(404, @body)
-      end)
+    test "handles a 404 response" do
+      {:error, response} = API.request(:get, "/404", %{}, [], [])
+      assert response ==
+        APIError.new({404, %{"message" => "{\"message\":\"Not Found\"}"}})
+    end
 
-      {:error, response} = API.request(:get, url_for("/foo"), [{"foo", "bar"}], @body, [:with_body])
-      assert response == APIError.new({404, %{"message" => @body}})
+    test "handles a 400 response" do
+      {:error, response} = API.request(:get, "/400", %{}, [], [])
+      assert response == APIError.new({400, %{"message" => "baz"}})
     end
 
-    @generic_error %{"message" => "foo"} |> Poison.encode!
+    test "handles a decode error for a 400..599 response" do
+      {:error, response} = API.request(:get, "/400-bad", %{}, [], [])
 
-    test "handles a 400 response", %{bypass: bypass} do
-      Bypass.expect(bypass, "GET", "/foo", fn %Conn{req_headers: headers} = conn ->
-        assert {"foo", "bar"} in headers
-        conn |> Conn.resp(400, @generic_error)
-      end)
+      assert response == HTTPClientError.new([reason: :body_decoding_error])
+    end
 
-      {:error, response} = API.request(:get, url_for("/foo"), [{"foo", "bar"}], @body, [:with_body])
-      assert response == APIError.new({400, @generic_error |> Poison.decode!})
+    test "handles a client error" do
+      {:error, %HTTPClientError{reason: reason}} =
+        API.request(:get, "/error", %{}, [], [])
+
+      assert reason == "Intentional Test Error"
     end
+  end
 
-    test "handles a decode error for a 400..599 response", %{bypass: bypass} do
-      Bypass.expect(bypass, "GET", "/foo", fn %Conn{req_headers: headers} = conn ->
-        assert {"foo", "bar"} in headers
-        conn |> Conn.resp(400, "foo")
-      end)
+  describe "get_all/3" do
+    defmodule MockPaginationAPI do
+      def request(:head, "/one-page", _body, _headers, _opts) do
+        {:ok, %HTTPoison.Response{headers: [], status_code: 200}}
+      end
+      def request(:head, "/two-pages", _body, _headers, _opts) do
+        next = '<two-pages/?page=2>; rel="next"'
+        last = '<two-pages/?page=2>; rel="last"'
+
+        headers = [{"Link", [next, last] |> Enum.join(", ")}]
+        {:ok, %HTTPoison.Response{headers: headers, status_code: 200}}
```

Comment:
This is all we needed really for the pagination test. IMO this satisfies most of what I wanted in #1132. What do you think @begedin?